### PR TITLE
Different wait for cancelled tasks

### LINKED ISSUE #94989: [CI] CancellableTasksIT testChildrenTasksCancelledOnTimeout failing
**Build scan:**
https://gradle-enterprise.elastic.co/s/7qx3wpwitkdiq/tests/:server:internalClusterTest/org.elasticsearch.action.admin.cluster.node.tasks.CancellableTasksIT/testChildrenTasksCancelledOnTimeout

**Reproduction line:**
```
./gradlew ':server:internalClusterTest' --tests "org.elasticsearch.action.admin.cluster.node.tasks.CancellableTasksIT.testChildrenTasksCancelledOnTimeout" -Dtests.seed=9F784DDF2A59F914 -Dtests.locale=es-US -Dtests.timezone=Asia/Ujung_Pandang -Druntime.java=20
```

**Applicable branches:**
main

**Reproduces locally?:**
No

**Failure history:**
https://gradle-enterprise.elastic.co/scans/tests?tests.container=org.elasticsearch.action.admin.cluster.node.tasks.CancellableTasksIT&tests.test=testChildrenTasksCancelledOnTimeout

**Failure excerpt:**
```
java.lang.AssertionError: expected:<0> but was:<2>

  at __randomizedtesting.SeedInfo.seed([9F784DDF2A59F914:6F379BC970940468]:0)
  at org.junit.Assert.fail(Assert.java:88)
  at org.junit.Assert.failNotEquals(Assert.java:834)
  at org.junit.Assert.assertEquals(Assert.java:645)
  at org.junit.Assert.assertEquals(Assert.java:631)
  at org.elasticsearch.action.admin.cluster.node.tasks.CancellableTasksIT.lambda$testChildrenTasksCancelledOnTimeout$9(CancellableTasksIT.java:390)
  at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:1152)
  at org.elasticsearch.action.admin.cluster.node.tasks.CancellableTasksIT.testChildrenTasksCancelledOnTimeout(CancellableTasksIT.java:386)
  at jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
  at java.lang.reflect.Method.invoke(Method.java:578)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1758)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:946)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:982)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:996)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:48)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:45)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:843)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:490)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:955)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:840)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:891)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:902)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:38)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:47)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.lambda$forkTimeoutingTask$0(ThreadLeakControl.java:850)
  at java.lang.Thread.run(Thread.java:1623)

```

### PR DESCRIPTION
**Pull request by kingherc**
As @fcofdez had [cleverly remarked](https://github.com/elastic/elasticsearch/pull/92588/files#r1088721302), the wait in this test was too harsh in general. Simply waiting for the latches of the cancelled children to complete is also enough.

Also adding @pxsalehi as reviewer in case you can better (or faster) review it.

Fixes #94989

### CODE REVIEW

File: server/src/internalClusterTest/java/org/elasticsearch/action/admin/cluster/node/tasks/CancellableTasksIT.java [hunk 1]
```diff
@@ -383,16 +382,12 @@ public void testChildrenTasksCancelledOnTimeout() throws Exception {
...
-                    assertEquals(0, stat.getActive());
-                    assertEquals(0, stat.getQueue());
-                }
-            }
-        }, 60L, TimeUnit.SECONDS);
         ensureBansAndCancellationsConsistency();
+
+        // Make sure all descendent requests have completed
+        for (TestRequest subRequest : rootRequest.descendants()) {
+            assertTrue(completedLatches.get(subRequest).await(60, TimeUnit.SECONDS));
```
Comment (DaveCTurner) [inline server/src/internalClusterTest/java/org/elasticsearch/action/admin/cluster/node/tasks/CancellableTasksIT.java:389]: 60s seems like a long wait per latch in a test. Could we use `ESTestCase#safeAwait(CountDownLatch)` which has a 10s timeout? (Also means there's no checked exceptions here too)

Comment (kingherc) [reply server/src/internalClusterTest/java/org/elasticsearch/action/admin/cluster/node/tasks/CancellableTasksIT.java]: @DaveCTurner I did not know about this one, thanks! Done.

**Comment by kingherc** [comment]
@elasticmachine please run elasticsearch-ci/bwc

File: server/src/internalClusterTest/java/org/elasticsearch/action/admin/cluster/node/tasks/CancellableTasksIT.java [hunk 2]
```diff
@@ -383,16 +381,12 @@ public void testChildrenTasksCancelledOnTimeout() throws Exception {
...
-                    assertEquals(0, stat.getQueue());
-                }
-            }
-        }, 60L, TimeUnit.SECONDS);
         ensureBansAndCancellationsConsistency();
+
+        // Make sure all descendent requests have completed
+        for (TestRequest subRequest : rootRequest.descendants()) {
+            safeAwait(completedLatches.get(subRequest));
+        }
```
[the remark this answers is not in the document]
Comment (kingherc) [reply server/src/internalClusterTest/java/org/elasticsearch/action/admin/cluster/node/tasks/CancellableTasksIT.java:389]: Hi! I do not think it matters. They both kind of depend on cancellations having finished (and thus both have timed waits), but they check different stuff.

**Comment by kingherc** [comment]
Hi, the latest CI revealed the basic bug of the test: I had made the cancellable task spinning, which in some cases resulted in all the generic thread pool being saturated, and not allowing the new cancellation requests to be done. I could not easily reproduce it on my laptop because I have many cores (I just caught it once), but it happens more regularly in the CI it seems. So I finally amended the PR to make the tasks not spinning: they are simply re-scheduled every 10ms until cancelled. This leaves "space" for the cancellation requests to be executed.

Feel free to review. cc @DaveCTurner , @pxsalehi

**Review by pxsalehi — APPROVED** [review]

**Comment by kingherc** [comment]
Thanks!

**Pull request merged**